content_sync/backends/base.py

```diff
@@ -0,0 +1,132 @@
+""" Sync abstract base """
+import abc
+from typing import Any
+
+from content_sync.models import ContentSyncState
+from websites.models import Website
+
+
+class BaseSyncBackend(abc.ABC):
+    """ Base class for syncing backends """
+
+    # NOTE: concrete implementations of this will probably want to
+    #       initialize a client object for the backing service (e.g. ghapi)
+    def __init__(self, website: Website):
+        self.website = website
+
+    @abc.abstractmethod
+    def create_website_in_backend(self):  # pragma: no cover
+        """
+        Called to create the website in the backend.
+
+        An example would be creating a VCS repository.
+        """
+        ...
+
+    @abc.abstractmethod
+    def create_backend_preview(self):  # pragma: no cover
+        """
+        Called to create a site preview in the backend.
+
+        An example would be merging the latest commits into a VCS' preview branch.
+        """
+        ...
+
+    @abc.abstractmethod
+    def create_backend_release(self):  # pragma: no cover
+        """
+        Called to create a site release in the backend.
+
+        An example would be merging the latest commits into a VCS' release branch.
+        """
+        ...
+
+    # NOTE: these next two could perform the same operations if the backing service
+    #       supports an UPSERT type operation (GitHub does)
+    @abc.abstractmethod
+    def create_content_in_backend(
+        self, sync_state: ContentSyncState
+    ):  # pragma: no cover
+        """
+        Called to create a piece of content in the backend.
+
+        An example would be commiting the contents of the WebsiteContent to the VCS repository.
+        """
+        ...
+
+    @abc.abstractmethod
+    def update_content_in_backend(
+        self, sync_state: ContentSyncState
+    ):  # pragma: no cover
+        """
+        Called to create a piece of content in the backend.
+
+        An example would be commiting the updated contents of the WebsiteContent to the VCS repository.
+        """
+        ...
+
+    @abc.abstractmethod
+    def delete_content_in_backend(
+        self, sync_state: ContentSyncState
+    ):  # pragma: no cover
+        """
+        Called to delete a piece of content in the backend.
+
+        An example would be making a commit to delete content in the VCS repository.
+        """
+        ...
+
+    # NOTE: this will be fired for every piece of content that is created, changed, or deleted in the database
```

Comment:
Fired how/where? A django signal if a particular backend has been specified in settings?

Reply:
I implemented a signal for create/update, will deal with delete separately since it's a little more challenging with the async behavior